File: iOS/BCFerries/HolidayRepository.swift
//
//  HolidayRepository.swift
//  BCFerries
//

import Foundation

open class HolidayRepository {
    private let holidayDao = HolidayDao()

    // Canada Day is a statutory holiday every year: if it's missing we have no holidays for this year
    private var canadaDay: LocalDate {
        LocalDate.parse("\(LocalDate.now().year!)-07-01")!
    }

    open func getHoliday(day: LocalDate) -> Bool {
        return holidayDao.query(today: day)
    }

    open func setHoliday(day: LocalDate, isHoliday: Bool) {
        if isHoliday {
            holidayDao.insert(date: day)
        } else {
            holidayDao.delete(date: day)
        }
    }

    open func shouldRefresh() -> Bool {
        return !getHoliday(day: canadaDay)
    }

    open func refresh(completion: @escaping () -> Void) {
        BCHolidayFetcher.fetch { holidays, _ in
            if holidays.count >= 10 {
                self.holidayDao.save(holidays: holidays)
            }
            completion()
        }
    }

}

File: Android/app/src/main/java/com/lunesu/bcferries/HolidayRepository.kt
package com.lunesu.bcferries

import org.joda.time.LocalDate

open class HolidayRepository(db: DbOpenHelper) {
    companion object {
        // Canada Day is a statutory holiday every year: if it's missing we have no holidays for this year
        private val CANADA_DAY: LocalDate get() = LocalDate(LocalDate.now().year, 7, 1)
    }

    private val holidayDao = HolidayDao(db)

    open fun getHoliday(day: LocalDate): Boolean {
        return holidayDao.query(day)
    }

    open fun setHoliday(day: LocalDate, isHoliday: Boolean) {
        if (isHoliday) {
            holidayDao.insert(day)
        } else {
            holidayDao.delete(day)
        }
    }

    open fun shouldRefresh(): Boolean {
        return !getHoliday(CANADA_DAY)
    }

    open suspend fun refresh() {
        runCatching {
            holidayDao.save(Utils.atLeast(BCHolidayFetcher.fetch(), 10))
        }
    }
}
